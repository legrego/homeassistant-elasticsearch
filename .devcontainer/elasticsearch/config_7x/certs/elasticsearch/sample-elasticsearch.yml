#
# SAMPLE ELASTICSEARCH CONFIGURATION FOR ENABLING SSL ON THE HTTP INTERFACE
#
# This is a sample configuration snippet for Elasticsearch that enables and configures SSL for the HTTP (Rest) interface
#
# This was automatically generated at: 2020-12-21 13:19:13-05:00
# This configuration was intended for Elasticsearch version 8.0.0
#
# You should review these settings, and then update the main configuration file at
#   /Users/larry/repos/kibana/.es/8.0.0/config/elasticsearch.yml
#

# This turns on SSL for the HTTP (Rest) interface
xpack.security.http.ssl.enabled: true

# This configures the keystore to use for SSL on HTTP
xpack.security.http.ssl.keystore.path: "http.p12"

