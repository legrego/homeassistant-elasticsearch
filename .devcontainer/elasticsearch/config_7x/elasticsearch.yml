#
# Path to directory where to store the data (separate multiple locations by comma):
#
path.data: /var/lib/elasticsearch
#
# Path to log files:
#
path.logs: /var/log/elasticsearch


cluster.name: home-assistant-cluster

xpack.security.http.ssl.enabled: true
xpack.security.http.ssl.keystore.path: certs/elasticsearch/http.p12
xpack.security.enabled: true