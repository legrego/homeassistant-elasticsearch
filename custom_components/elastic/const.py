""" constants """

DOMAIN = 'elastic'

CONF_PUBLISH_ENABLED = 'publish_enabled'
CONF_HEALTH_SENSOR_ENABLED = 'health_sensor_enabled'
CONF_INDEX_FORMAT = 'index_format'
CONF_PUBLISH_FREQUENCY = 'publish_frequency'
CONF_ONLY_PUBLISH_CHANGED = 'only_publish_changed'
CONF_ILM_ENABLED = 'ilm_enabled'
CONF_ILM_POLICY_NAME = 'ilm_policy_name'
CONF_ILM_MAX_SIZE = 'ilm_max_size'
CONF_ILM_DELETE_AFTER = 'ilm_delete_after'
CONF_SSL_CA_PATH = 'ssl_ca_path'
CONF_CLOUD_ID = 'cloud_id'

CONF_TAGS = 'tags'

ONE_MINUTE = 60
ONE_HOUR = 60 * 60

VERSION_SUFFIX = "-v4_1"

INDEX_TEMPLATE_NAME = "hass-index-template" + VERSION_SUFFIX
