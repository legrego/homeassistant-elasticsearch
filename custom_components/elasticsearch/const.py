""" constants """

DOMAIN = "elasticsearch"

CONF_PUBLISH_ENABLED = "publish_enabled"
CONF_HEALTH_SENSOR_ENABLED = "health_sensor_enabled"
CONF_INDEX_FORMAT = "index_format"
CONF_PUBLISH_FREQUENCY = "publish_frequency"
CONF_EXCLUDED_DOMAINS = "excluded_domains"
CONF_EXCLUDED_ENTITIES = "excluded_entities"
CONF_PUBLISH_MODE = "publish_mode"
CONF_INCLUDED_DOMAINS = "included_domains"
CONF_INCLUDED_ENTITIES = "included_entities"
CONF_ONLY_PUBLISH_CHANGED = "only_publish_changed"
CONF_ILM_ENABLED = "ilm_enabled"
CONF_ILM_POLICY_NAME = "ilm_policy_name"
CONF_ILM_MAX_SIZE = "ilm_max_size"
CONF_ILM_DELETE_AFTER = "ilm_delete_after"
CONF_SSL_CA_PATH = "ssl_ca_path"

CONF_TAGS = "tags"

ONE_MINUTE = 60
ONE_HOUR = 60 * 60

VERSION_SUFFIX = "-v4_1"

INDEX_TEMPLATE_NAME = "hass-index-template" + VERSION_SUFFIX

PUBLISH_MODE_ALL = "All"
PUBLISH_MODE_STATE_CHANGES = "State changes"
PUBLISH_MODE_ANY_CHANGES = "Any changes"
