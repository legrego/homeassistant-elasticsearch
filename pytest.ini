[pytest]
addopts = -rxf -x -v -l --cov=./ --cov-report=xml
filterwarnings =
    ignore::DeprecationWarning
testpaths = tests
python_files =
    test_*.py