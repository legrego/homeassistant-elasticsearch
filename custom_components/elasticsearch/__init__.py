"""
Support for sending event data to an Elasticsearch cluster
"""

from copy import deepcopy

import homeassistant.helpers.config_validation as cv
import voluptuous as vol
from homeassistant.config_entries import SOURCE_IMPORT, ConfigEntry
from homeassistant.const import (
    CONF_ALIAS,
    CONF_DOMAINS,
    CONF_ENTITIES,
    CONF_EXCLUDE,
    CONF_PASSWORD,
    CONF_TIMEOUT,
    CONF_URL,
    CONF_USERNAME,
    CONF_VERIFY_SSL,
)
from homeassistant.exceptions import ConfigEntryAuthFailed, ConfigEntryNotReady
from homeassistant.helpers.typing import HomeAssistantType

from custom_components.elasticsearch.utils import get_merged_config

from .const import (
    CONF_EXCLUDED_DOMAINS,
    CONF_EXCLUDED_ENTITIES,
    CONF_HEALTH_SENSOR_ENABLED,
    CONF_ILM_DELETE_AFTER,
    CONF_ILM_ENABLED,
    CONF_ILM_MAX_SIZE,
    CONF_ILM_POLICY_NAME,
    CONF_INDEX_FORMAT,
    CONF_ONLY_PUBLISH_CHANGED,
    CONF_PUBLISH_ENABLED,
    CONF_PUBLISH_FREQUENCY,
    CONF_PUBLISH_MODE,
    CONF_SSL_CA_PATH,
    CONF_TAGS,
    DOMAIN,
    ONE_MINUTE,
    PUBLISH_MODE_ALL,
    PUBLISH_MODE_ANY_CHANGES,
)
from .errors import AuthenticationRequired, InsufficientPrivileges, UnsupportedVersion
from .es_integration import ElasticIntegration
from .logger import LOGGER

# Legacy (yml-based) configuration schema
CONFIG_SCHEMA = vol.Schema(
    {
        DOMAIN: vol.Schema(
            {
                vol.Optional(CONF_URL): cv.url,
                vol.Optional(CONF_USERNAME): cv.string,
                vol.Optional(CONF_PASSWORD): cv.string,
                vol.Optional(CONF_TIMEOUT, default=30): cv.positive_int,
                vol.Optional(CONF_PUBLISH_ENABLED, default=True): cv.boolean,
                vol.Optional(CONF_HEALTH_SENSOR_ENABLED, default=True): cv.boolean,
                vol.Optional(CONF_PASSWORD): cv.string,
                vol.Optional(CONF_INDEX_FORMAT, default="hass-events"): cv.string,
                vol.Optional(CONF_ALIAS, default="active-hass-index"): cv.string,
                vol.Optional(
                    CONF_PUBLISH_FREQUENCY, default=ONE_MINUTE
                ): cv.positive_int,
                vol.Optional(CONF_ONLY_PUBLISH_CHANGED, default=False): cv.boolean,
                vol.Optional(CONF_ILM_ENABLED, default=True): cv.boolean,
                vol.Optional(CONF_ILM_POLICY_NAME, default="home-assistant"): cv.string,
                vol.Optional(CONF_ILM_MAX_SIZE, default="30gb"): cv.string,
                vol.Optional(CONF_ILM_DELETE_AFTER, default="365d"): cv.string,
                vol.Optional(CONF_VERIFY_SSL, default=True): cv.boolean,
                vol.Optional(CONF_SSL_CA_PATH): cv.string,
                vol.Optional(CONF_TAGS, default=["hass"]): vol.All(
                    cv.ensure_list, [cv.string]
                ),
                vol.Optional(CONF_EXCLUDED_DOMAINS, default=[]): vol.All(
                    cv.ensure_list, [cv.string]
                ),
                vol.Optional(CONF_EXCLUDED_ENTITIES, default=[]): cv.entity_ids,
                vol.Optional(CONF_EXCLUDE, default={}): vol.Schema(
                    {
                        vol.Optional(CONF_DOMAINS, default=[]): vol.All(
                            cv.ensure_list, [cv.string]
                        ),
                        vol.Optional(CONF_ENTITIES, default=[]): cv.entity_ids,
                    }
                ),
            }
        ),
    },
    extra=vol.ALLOW_EXTRA,
)


async def async_setup(hass: HomeAssistantType, config):
    """Set up Elasticsearch integration via legacy yml-based setup."""
    if DOMAIN not in config:
        return True

    conf = config[DOMAIN]

    # Migrate legacy yml-based config to a flattened structure.
    excluded = conf.get(CONF_EXCLUDE, {})
    conf[CONF_EXCLUDED_DOMAINS] = excluded.get(CONF_DOMAINS, [])
    conf[CONF_EXCLUDED_ENTITIES] = excluded.get(CONF_ENTITIES, [])

    # Run legacy yml-based config through the config flow.
    hass.async_create_task(
        hass.config_entries.flow.async_init(
            DOMAIN, context={"source": SOURCE_IMPORT}, data=deepcopy(conf)
        )
    )

    return True


async def async_migrate_entry(hass, config_entry: ConfigEntry):
    """Migrate old entry."""
    LOGGER.debug("Migrating config entry from version %s", config_entry.version)

    if config_entry.version == 1:

        new = get_merged_config(config_entry)

        only_publish_changed = new.get(CONF_ONLY_PUBLISH_CHANGED, False)
        new[CONF_PUBLISH_MODE] = (
            PUBLISH_MODE_ALL if not only_publish_changed else PUBLISH_MODE_ANY_CHANGES
        )

        if CONF_ONLY_PUBLISH_CHANGED in new:
            del new[CONF_ONLY_PUBLISH_CHANGED]

        config_entry.data = {**new}

        config_entry.version = 2

    LOGGER.info("Migration to version %s successful", config_entry.version)

    return True


async def async_setup_entry(hass: HomeAssistantType, config_entry: ConfigEntry):
    """ Setup integration via config flow. """

    LOGGER.debug("Setting up integtation")
    init = await _async_init_integration(hass, config_entry)
    config_entry.add_update_listener(async_config_entry_updated)
    return init


async def async_unload_entry(hass: HomeAssistantType, config_entry: ConfigEntry):
    """ Teardown integration. """
    existing_instance = hass.data.get(DOMAIN)
    if isinstance(existing_instance, ElasticIntegration):
        LOGGER.debug("Shutting down previous integration")
        await existing_instance.async_shutdown(config_entry)
        hass.data[DOMAIN] = None
    return True


async def async_config_entry_updated(
    hass: HomeAssistantType, config_entry: ConfigEntry
):
    """ Respond to config changes. """
    LOGGER.debug("Configuration change detected")
    return await _async_init_integration(hass, config_entry)


async def _async_init_integration(hass: HomeAssistantType, config_entry: ConfigEntry):
    """ Initialize integration. """
    await async_unload_entry(hass, config_entry)

    integration = None
    try:
        integration = ElasticIntegration(hass, config_entry)
        await integration.async_init()
    except UnsupportedVersion as err:
        LOGGER.error("Unsupported Elasticsearch version detected")
        raise ConfigEntryNotReady from err
    except AuthenticationRequired as err:
        LOGGER.error("Missing or invalid credentials")
        raise ConfigEntryAuthFailed from err
    except InsufficientPrivileges as err:
        LOGGER.error("Account does not have sufficient privileges")
        raise ConfigEntryAuthFailed from err

    hass.data[DOMAIN] = integration

    return True
