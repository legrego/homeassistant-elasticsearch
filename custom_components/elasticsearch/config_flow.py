"""Config flow for Elastic."""


import homeassistant.helpers.config_validation as cv
import voluptuous as vol
from homeassistant import config_entries
from homeassistant.config_entries import SOURCE_IGNORE, SOURCE_IMPORT
from homeassistant.const import (
    CONF_ALIAS,
    CONF_PASSWORD,
    CONF_TIMEOUT,
    CONF_URL,
    CONF_USERNAME,
    CONF_VERIFY_SSL,
)
from homeassistant.core import callback
from homeassistant.helpers.selector import selector

from .const import (
    CONF_EXCLUDED_DOMAINS,
    CONF_EXCLUDED_ENTITIES,
    CONF_HEALTH_SENSOR_ENABLED,
    CONF_ILM_DELETE_AFTER,
    CONF_ILM_ENABLED,
    CONF_ILM_MAX_SIZE,
    CONF_ILM_POLICY_NAME,
    CONF_INCLUDED_DOMAINS,
    CONF_INCLUDED_ENTITIES,
    CONF_INDEX_FORMAT,
    CONF_PUBLISH_ENABLED,
    CONF_PUBLISH_FREQUENCY,
    CONF_PUBLISH_MODE,
    CONF_SSL_CA_PATH,
)
from .const import DOMAIN as ELASTIC_DOMAIN
from .const import (
    ONE_MINUTE,
    PUBLISH_MODE_ALL,
    PUBLISH_MODE_ANY_CHANGES,
    PUBLISH_MODE_STATE_CHANGES,
)
from .errors import (
    AuthenticationRequired,
    CannotConnect,
    InsufficientPrivileges,
    UnsupportedVersion,
    UntrustedCertificate,
)
from .es_gateway import ElasticsearchGateway
from .logger import LOGGER

DEFAULT_URL = "http://localhost:9200"
DEFAULT_ALIAS = "active-hass-index"
DEFAULT_INDEX_FORMAT = "hass-events"
DEFAULT_PUBLISH_ENABLED = True
DEFAULT_PUBLISH_FREQUENCY = ONE_MINUTE
DEFAULT_PUBLISH_MODE = PUBLISH_MODE_ANY_CHANGES
DEFAULT_VERIFY_SSL = True
DEFAULT_TIMEOUT_SECONDS = 30
DEFAULT_ILM_ENABLED = True
DEFAULT_ILM_POLICY_NAME = "home-assistant"
DEFAULT_ILM_MAX_SIZE = "30gb"
DEFAULT_ILM_DELETE_AFTER = "365d"


class ElasticFlowHandler(config_entries.ConfigFlow, domain=ELASTIC_DOMAIN):
    """Handle an Elastic config flow."""

    VERSION = 2
    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_PUSH

    @staticmethod
    @callback
    def async_get_options_flow(config_entry):
        """Get the options flow for this handler."""
        return ElasticOptionsFlowHandler(config_entry)

    def __init__(self):
        """Initialize the Elastic flow."""
        self.config = {}

        self.tls_schema = {
            vol.Optional(CONF_VERIFY_SSL, default=DEFAULT_VERIFY_SSL): bool,
            vol.Optional(CONF_SSL_CA_PATH, default=""): str,
        }

    def build_setup_schema(self):
        schema = {
            vol.Required(
                CONF_URL, default=self.config.get(CONF_URL, "http://localhost:9200")
            ): str,
            vol.Optional(CONF_USERNAME): str,
            vol.Optional(CONF_PASSWORD): str,
        }

        if self.show_advanced_options:
            schema[
                vol.Required(
                    CONF_TIMEOUT,
                    default=self.config.get(CONF_TIMEOUT, DEFAULT_TIMEOUT_SECONDS),
                )
            ] = int

        return schema

    def build_full_config(self, user_input={}):
        return {
            CONF_URL: user_input.get(CONF_URL, DEFAULT_URL),
            CONF_USERNAME: user_input.get(CONF_USERNAME),
            CONF_PASSWORD: user_input.get(CONF_PASSWORD),
            CONF_TIMEOUT: user_input.get(CONF_TIMEOUT, DEFAULT_TIMEOUT_SECONDS),
            CONF_VERIFY_SSL: user_input.get(CONF_VERIFY_SSL, DEFAULT_VERIFY_SSL),
            CONF_SSL_CA_PATH: user_input.get(CONF_SSL_CA_PATH, None),
            CONF_PUBLISH_ENABLED: user_input.get(
                CONF_PUBLISH_ENABLED, DEFAULT_PUBLISH_ENABLED
            ),
            CONF_PUBLISH_FREQUENCY: user_input.get(
                CONF_PUBLISH_FREQUENCY, DEFAULT_PUBLISH_FREQUENCY
            ),
            CONF_PUBLISH_MODE: user_input.get(CONF_PUBLISH_MODE, DEFAULT_PUBLISH_MODE),
            CONF_ALIAS: user_input.get(CONF_ALIAS, DEFAULT_ALIAS),
            CONF_INDEX_FORMAT: user_input.get(CONF_INDEX_FORMAT, DEFAULT_INDEX_FORMAT),
            CONF_EXCLUDED_DOMAINS: user_input.get(CONF_EXCLUDED_DOMAINS, []),
            CONF_EXCLUDED_ENTITIES: user_input.get(CONF_EXCLUDED_ENTITIES, []),
            CONF_INCLUDED_DOMAINS: user_input.get(CONF_INCLUDED_DOMAINS, []),
            CONF_INCLUDED_ENTITIES: user_input.get(CONF_INCLUDED_ENTITIES, []),
            CONF_HEALTH_SENSOR_ENABLED: user_input.get(
                CONF_HEALTH_SENSOR_ENABLED, True
            ),
            CONF_ILM_ENABLED: user_input.get(CONF_ILM_ENABLED, DEFAULT_ILM_ENABLED),
            CONF_ILM_POLICY_NAME: user_input.get(
                CONF_ILM_POLICY_NAME, DEFAULT_ILM_POLICY_NAME
            ),
            CONF_ILM_MAX_SIZE: user_input.get(CONF_ILM_MAX_SIZE, DEFAULT_ILM_MAX_SIZE),
            CONF_ILM_DELETE_AFTER: user_input.get(
                CONF_ILM_DELETE_AFTER, DEFAULT_ILM_DELETE_AFTER
            ),
        }

    async def async_step_user(self, user_input=None):
        """Handle a flow initialized by the user."""
        if self._async_current_entries():
            return self.async_abort(reason="single_instance_allowed")

        if user_input is None:
            return self.async_show_form(
                step_id="user",
                data_schema=vol.Schema(self.build_setup_schema()),
            )

        self.config = self.build_full_config(user_input)

        return await self._async_elasticsearch_login()

    async def async_step_tls(self, user_input=None):
        """Handle establishing a trusted connection to Elasticsearch."""

        if user_input is None:
            return self.async_show_form(
                step_id="tls", data_schema=vol.Schema(self.tls_schema)
            )

        self.config[CONF_VERIFY_SSL] = user_input[CONF_VERIFY_SSL]
        if len(user_input[CONF_SSL_CA_PATH]):
            self.config[CONF_SSL_CA_PATH] = user_input[CONF_SSL_CA_PATH]

        return await self._async_elasticsearch_login()

    async def async_step_import(self, import_config):
        """Import a config entry from configuration.yaml."""
        # Check if new config entry matches any existing config entries
        entries = self.hass.config_entries.async_entries(ELASTIC_DOMAIN)
        for entry in entries:
            # If source is ignore bypass host check and continue through loop
            if entry.source == SOURCE_IGNORE:
                continue

            if entry.data[CONF_URL] == import_config[CONF_URL]:
                self.hass.config_entries.async_update_entry(
                    entry=entry,
                    data=self.build_full_config(import_config),
                    options=import_config,
                )
                return self.async_abort(reason="updated_entry")

        if entries:
            LOGGER.warning("Already configured. Only a single configuration possible.")
            return self.async_abort(reason="single_instance_allowed")

        return await self.async_step_user(user_input=import_config)

    async def _async_elasticsearch_login(self):
        """Handle connection & authentication to Elasticsearch"""
        errors = {}

        try:
            gateway = ElasticsearchGateway(self.config)
            await gateway.check_connection(self.hass)
        except UntrustedCertificate:
            errors["base"] = "untrusted_connection"
            return self.async_show_form(
                step_id="tls", data_schema=vol.Schema(self.tls_schema), errors=errors
            )
        except AuthenticationRequired:
            errors["base"] = "invalid_auth"
        except InsufficientPrivileges:
            errors["base"] = "insufficient_privileges"
        except CannotConnect:
            errors["base"] = "cannot_connect"
        except UnsupportedVersion:
            errors["base"] = "unsupported_version"
        except Exception as ex:  # pylint: disable=broad-except
            LOGGER.error(
                "Unknown error connecting with Elasticsearch cluster. %s",
                ex,
            )
            errors["base"] = "cannot_connect"

        if errors:
            return self.async_show_form(
                step_id="user",
                data_schema=vol.Schema(self.build_setup_schema()),
                errors=errors,
            )

        return await self._async_create_entry()

    async def _async_create_entry(self):
        """Create the config entry."""
        existing_entry = await self.async_set_unique_id(self.config[CONF_URL])

        if existing_entry:
            self.hass.config_entries.async_update_entry(
                existing_entry, data=self.config
            )
            # Reload the config entry otherwise devices will remain unavailable
            self.hass.async_create_task(
                self.hass.config_entries.async_reload(existing_entry.entry_id)
            )

            return self.async_abort(reason="reauth_successful")

        return self.async_create_entry(title=self.config[CONF_URL], data=self.config)


class ElasticOptionsFlowHandler(config_entries.OptionsFlow):
    """Handle Elastic options."""

    def __init__(self, config_entry):
        """Initialize Elastic options flow."""
        self.config_entry = config_entry
        self.options = dict(config_entry.options)

    async def async_step_init(self, user_input=None):
        """Manage the Elastic options."""

        if self.config_entry.source == SOURCE_IMPORT:
            return await self.async_step_yaml(user_input)

        return await self.async_step_publish_options()

    async def async_step_yaml(self, user_input=None):
        """No options for yaml managed entries."""
        return self.async_abort(reason="configured_via_yaml")

    async def async_step_publish_options(self, user_input=None):
        """Publish Options"""
        if user_input is not None:
            self.options.update(user_input)
            return await self.async_step_ilm_options()

        return self.async_show_form(
            step_id="publish_options",
            data_schema=vol.Schema(await self.async_build_publish_options_schema()),
        )

    async def async_step_ilm_options(self, user_input=None):
        """ILM Options"""
        errors = {}

        if user_input is not None:
            self.options.update(user_input)
            return await self.async_step_health_options()

        return self.async_show_form(
            step_id="ilm_options",
            data_schema=vol.Schema(self._build_ilm_options_schema()),
            errors=errors,
        )

    async def async_step_health_options(self, user_input=None):
        """ Health Sensor Options"""
        if user_input is not None:
            self.options.update(user_input)
            return await self._update_options()

        return self.async_show_form(
            step_id="health_options",
            data_schema=vol.Schema(self._build_health_options_schema()),
        )

    async def _update_options(self):
        """Update config entry options."""
        return self.async_create_entry(title="", data=self.options)

    def _get_config_value(self, key, default):
        current = self.options.get(key, default)
        if current is None:
            return default
        return current

    async def async_build_publish_options_schema(self):
        """Builds the schema for publish options."""
        domains, entities = await self._async_get_domains_and_entities()

        schema = {
            vol.Required(
                CONF_PUBLISH_ENABLED,
                default=self._get_config_value(
                    CONF_PUBLISH_ENABLED, DEFAULT_PUBLISH_ENABLED
                ),
            ): bool,
            vol.Required(
                CONF_PUBLISH_FREQUENCY,
                default=self._get_config_value(
                    CONF_PUBLISH_FREQUENCY, DEFAULT_PUBLISH_FREQUENCY
                ),
            ): int,
            vol.Required(
                CONF_PUBLISH_MODE,
                default=self._get_config_value(CONF_PUBLISH_MODE, DEFAULT_PUBLISH_MODE),
            ): selector(
                {
                    "select": {
                        "options": [
                            {"label": "All entities", "value": PUBLISH_MODE_ALL},
                            {
                                "label": "Entities with state changes",
                                "value": PUBLISH_MODE_STATE_CHANGES,
                            },
                            {
                                "label": "Entities with state or attribute changes",
                                "value": PUBLISH_MODE_ANY_CHANGES,
                            },
                        ]
                    }
                }
            ),
            vol.Required(
                CONF_EXCLUDED_DOMAINS,
                default=self._get_config_value(CONF_EXCLUDED_DOMAINS, []),
            ): cv.multi_select(domains),
            vol.Required(
                CONF_EXCLUDED_ENTITIES,
                default=self._get_config_value(CONF_EXCLUDED_ENTITIES, []),
            ): cv.multi_select(entities),
            vol.Required(
                CONF_INCLUDED_DOMAINS,
                default=self._get_config_value(CONF_INCLUDED_DOMAINS, []),
            ): cv.multi_select(domains),
            vol.Required(
                CONF_INCLUDED_ENTITIES,
                default=self._get_config_value(CONF_INCLUDED_ENTITIES, []),
            ): cv.multi_select(entities),
        }

        if self.show_advanced_options:
            schema[
                vol.Required(
                    CONF_INDEX_FORMAT,
                    default=self._get_config_value(
                        CONF_INDEX_FORMAT, DEFAULT_INDEX_FORMAT
                    ),
                )
            ] = str

            schema[
                vol.Required(
                    CONF_ALIAS,
                    default=self._get_config_value(CONF_ALIAS, DEFAULT_ALIAS),
                )
            ] = str

        return schema

    def _build_ilm_options_schema(self):
        schema = {
            vol.Required(
                CONF_ILM_ENABLED, default=self._get_config_value(CONF_ILM_ENABLED, True)
            ): bool,
            vol.Required(
                CONF_ILM_POLICY_NAME,
                default=self._get_config_value(
                    CONF_ILM_POLICY_NAME, DEFAULT_ILM_POLICY_NAME
                ),
            ): str,
            vol.Required(
                CONF_ILM_MAX_SIZE,
                default=self._get_config_value(CONF_ILM_MAX_SIZE, DEFAULT_ILM_MAX_SIZE),
            ): str,
            vol.Required(
                CONF_ILM_DELETE_AFTER,
                default=self._get_config_value(
                    CONF_ILM_DELETE_AFTER, DEFAULT_ILM_DELETE_AFTER
                ),
            ): str,
        }

        return schema

    def _build_health_options_schema(self):
        schema = {
            vol.Required(
                CONF_HEALTH_SENSOR_ENABLED,
                default=self._get_config_value(CONF_HEALTH_SENSOR_ENABLED, True),
            ): bool,
        }

        return schema

    @callback
    async def _async_get_domains_and_entities(self):
        states = self.hass.states.async_all()
        domains = set()
        entity_ids = []

        for state in states:
            entity_ids.append(state.entity_id)
            domains.add(state.domain)

        return sorted(list(domains)), sorted(entity_ids)
