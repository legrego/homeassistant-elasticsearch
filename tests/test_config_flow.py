from unittest.mock import MagicMock

import aiohttp
import pytest
from homeassistant import data_entry_flow
from homeassistant.config_entries import SOURCE_USER
from homeassistant.helpers.typing import HomeAssistantType

from custom_components.elasticsearch.const import DOMAIN
from tests.test_util.es_startup_mocks import mock_es_initialization


async def test_user_flow_minimum_fields(hass: HomeAssistantType, aioclient_mock):
    """ Test user config flow with minimum fields. """

    result = await hass.config_entries.flow.async_init(
        DOMAIN, context={"source": SOURCE_USER}
    )
    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
    assert result["step_id"] == "user"

    es_url = "http://minimum-fields:9200"

    mock_es_initialization(
        aioclient_mock,
        url=es_url,
        mock_health_check=True,
        mock_index_creation=True,
        mock_template_setup=True,
    )

    result = await hass.config_entries.flow.async_configure(
        result["flow_id"], user_input={"url": es_url}
    )

    assert result["type"] == data_entry_flow.RESULT_TYPE_CREATE_ENTRY
    assert result["title"] == es_url
    assert result["data"]["url"] == es_url
    assert result["data"]["username"] is None
    assert result["data"]["password"] is None
    assert result["data"]["ssl_ca_path"] is None
    assert result["data"]["verify_ssl"] is True
    assert result["data"]["publish_enabled"] is True
    assert result["data"]["health_sensor_enabled"] is True


async def test_user_flow_unsupported_version(hass: HomeAssistantType, aioclient_mock):
    """ Test user config flow with minimum fields. """

    result = await hass.config_entries.flow.async_init(
        DOMAIN, context={"source": SOURCE_USER}
    )
    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
    assert result["step_id"] == "user"

    es_url = "http://minimum-fields:9200"

    mock_es_initialization(aioclient_mock, url=es_url, mock_unsupported_version=True)

    result = await hass.config_entries.flow.async_configure(
        result["flow_id"], user_input={"url": es_url}
    )

    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
    assert result["step_id"] == "user"
    assert result["errors"]["base"] == "unsupported_version"


async def test_user_flow_to_tls_flow(hass: HomeAssistantType, aioclient_mock):
    """ Test user config flow with config that forces TLS configuration. """
    result = await hass.config_entries.flow.async_init(
        DOMAIN, context={"source": SOURCE_USER}
    )
    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
    assert result["step_id"] == "user"

    es_url = "https://untrusted-connection:9200"

    class MockSSLError(aiohttp.client_exceptions.ClientConnectorCertificateError):
        """
        Mocks an SSL error caused by an untrusted certificate.
        This is imperfect, but gets the job done for now.
        """

        def __init__(self):
            self._conn_key = MagicMock()
            self._certificate_error = Exception("AHHHH")
            return

    aioclient_mock.get(es_url, exc=MockSSLError)

    result = await hass.config_entries.flow.async_configure(
        result["flow_id"], user_input={"url": es_url}
    )

    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
    assert result["errors"]["base"] == "untrusted_connection"
    assert result["step_id"] == "tls"
    assert "data" not in result


async def test_flow_fails_es_unavailable(hass: HomeAssistantType, aioclient_mock):
    """ Test user config flow fails if connection cannot be established. """
    result = await hass.config_entries.flow.async_init(
        DOMAIN, context={"source": SOURCE_USER}
    )
    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
    assert result["step_id"] == "user"

    es_url = "http://unavailable-host:9200"

    aioclient_mock.get(es_url, exc=aiohttp.ClientError)

    result = await hass.config_entries.flow.async_configure(
        result["flow_id"], user_input={"url": es_url}
    )

    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
    assert result["errors"]["base"] == "cannot_connect"
    assert result["step_id"] == "user"
    assert "data" not in result


async def test_flow_fails_unauthorized(hass: HomeAssistantType, aioclient_mock):
    """ Test user config flow fails if connection cannot be established. """
    result = await hass.config_entries.flow.async_init(
        DOMAIN, context={"source": SOURCE_USER}
    )
    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
    assert result["step_id"] == "user"

    es_url = "http://needs-auth:9200"

    aioclient_mock.get(es_url, status=401)

    result = await hass.config_entries.flow.async_configure(
        result["flow_id"], user_input={"url": es_url}
    )

    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
    assert result["errors"]["base"] == "invalid_auth"
    assert result["step_id"] == "user"
    assert "data" not in result


@pytest.mark.skip(
    reason="Struggling to prevent real network requests. Something is not honoring our aioclient mock."
)
async def test_options_flow(
    hass: HomeAssistantType, aioclient_mock, event_loop
) -> None:
    """Test options config flow."""

    es_url = "http://localhost:9200"

    mock_es_initialization(
        aioclient_mock,
        url=es_url,
        mock_health_check=True,
        mock_index_creation=True,
        mock_template_setup=True,
        mock_ilm_setup=True,
    )

    result = await hass.config_entries.flow.async_init(
        DOMAIN, context={"source": SOURCE_USER}, data={"url": es_url}
    )
    await hass.async_block_till_done()
    assert result["type"] == data_entry_flow.RESULT_TYPE_CREATE_ENTRY
    entry = result["result"]

    result = await hass.config_entries.options.async_init(entry.entry_id, data=None)

    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
    assert result["step_id"] == "publish_options"

    result = await hass.config_entries.options.async_configure(
        result["flow_id"], user_input={}
    )

    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
    assert result["step_id"] == "ilm_options"

    result = await hass.config_entries.options.async_configure(
        result["flow_id"], user_input={}
    )

    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
    assert result["step_id"] == "health_options"

    result = await hass.config_entries.options.async_configure(
        result["flow_id"], user_input={}
    )

    assert result["type"] == data_entry_flow.RESULT_TYPE_CREATE_ENTRY
