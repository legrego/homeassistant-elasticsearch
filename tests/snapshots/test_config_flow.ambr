# serializer version: 1
# name: Test_Integration_Tests.test_basic_auth_flow
  dict({
    'data': dict({
      'password': 'changeme',
      'ssl_ca_path': None,
      'timeout': 30,
      'url': 'http://basic-auth-flow:9200',
      'username': 'hass_writer',
      'verify_ssl': True,
    }),
    'options': dict({
      'change_detection_enabled': True,
      'change_detection_type': list([
        'STATE',
        'ATTRIBUTE',
      ]),
      'excluded_domains': list([
      ]),
      'excluded_entities': list([
      ]),
      'included_domains': list([
      ]),
      'included_entities': list([
      ]),
      'polling_frequency': 60,
      'publish_frequency': 60,
    }),
  })
# ---
# name: Test_Integration_Tests.test_no_auth_flow_isolate
  dict({
    'data': dict({
      'ssl_ca_path': None,
      'timeout': 30,
      'url': 'http://minimum-fields:9200',
      'verify_ssl': True,
    }),
    'options': dict({
      'change_detection_enabled': True,
      'change_detection_type': list([
        'STATE',
        'ATTRIBUTE',
      ]),
      'excluded_domains': list([
      ]),
      'excluded_entities': list([
      ]),
      'included_domains': list([
      ]),
      'included_entities': list([
      ]),
      'polling_frequency': 60,
      'publish_frequency': 60,
    }),
  })
# ---
